v2/elasticsearch-common/src/main/java/com/google/cloud/teleport/v2/elasticsearch/utils/ElasticsearchIO.java

```diff
@@ -0,0 +1,1508 @@
+          .setMaxAttempts(maxAttempts)
+          .setMaxDuration(maxDuration)
+          .setRetryPredicate(DEFAULT_RETRY_PREDICATE)
+          .build();
+    }
+
+    // Exposed only to allow tests to easily simulate server errors
+    @VisibleForTesting
+    RetryConfiguration withRetryPredicate(RetryPredicate predicate) {
+      checkArgument(predicate != null, "predicate must be provided");
+      return builder().setRetryPredicate(predicate).build();
+    }
+
+    /**
+     * An interface used to control if we retry the Elasticsearch call when a {@link Response} is
+     * obtained. If {@link RetryPredicate#test(Object)} returns true, {@link Write} tries to resend
+     * the requests to the Elasticsearch server if the {@link RetryConfiguration} permits it.
+     */
+    @FunctionalInterface
+    interface RetryPredicate extends Predicate<HttpEntity>, Serializable {}
+
+    /**
+     * This is the default predicate used to test if a failed ES operation should be retried. A
+     * retry will be attempted until the maxAttempts or maxDuration is exceeded, whichever comes
+     * first, for TOO_MANY_REQUESTS(429) error.
+     */
+    @VisibleForTesting
+    static class DefaultRetryPredicate implements RetryPredicate {
+
+      private int errorCode;
+
+      DefaultRetryPredicate(int code) {
+        this.errorCode = code;
+      }
+
+      DefaultRetryPredicate() {
+        this(429);
+      }
+
+      /** Returns true if the response has the error code for any mutation. */
+      private static boolean errorCodePresent(HttpEntity responseEntity, int errorCode) {
+        try {
+          JsonNode json = parseResponse(responseEntity);
+          if (json.path("errors").asBoolean()) {
+            for (JsonNode item : json.path("items")) {
+              if (item.findValue("status").asInt() == errorCode) {
+                return true;
+              }
+            }
+          }
+        } catch (IOException e) {
+          LOG.warn("Could not extract error codes from responseEntity {}", responseEntity);
+        }
+        return false;
+      }
+
+      @Override
+      public boolean test(HttpEntity responseEntity) {
+        return errorCodePresent(responseEntity, errorCode);
+      }
+    }
+  }
+
+  /** A {@link PTransform} writing data to Elasticsearch. */
+  @AutoValue
+  public abstract static class Write extends PTransform<PCollection<String>, PDone> {
+
+    /**
+     * Interface allowing a specific field value to be returned from a parsed JSON document. This is
+     * used for using explicit document ids, and for dynamic routing (index/Type) on a document
+     * basis. A null response will result in default behaviour and an exception will be propagated
+     * as a failure.
+     */
+    public interface FieldValueExtractFn extends SerializableFunction<JsonNode, String> {}
+
+    public interface BooleanFieldValueExtractFn extends SerializableFunction<JsonNode, Boolean> {}
```

Comment:
Include JavaDoc comments for BooleanFieldValueExtractFn to avoid checkstyle exception